src/execution/index/art/art.cpp

```diff
@@ -204,6 +204,160 @@ void ART::GenerateKeys(DataChunk &input, vector<unique_ptr<Key>> &keys) {
 	}
 }
 
+struct KeySection {
+	KeySection(idx_t start_p, idx_t end_p, idx_t depth_p, data_t key_byte_p)
+	    : start(start_p), end(end_p), depth(depth_p), key_byte(key_byte_p) {};
+	idx_t start;
+	idx_t end;
+	idx_t depth;
+	data_t key_byte;
+};
+
+void GetNodeChildren(vector<KeySection> &child_sections, vector<unique_ptr<Key>> &keys, KeySection &key_section) {
+
+	idx_t child_start_idx = key_section.start;
+	for (idx_t i = key_section.start + 1; i <= key_section.end; i++) {
+		if (keys[i - 1]->data[key_section.depth] != keys[i]->data[key_section.depth]) {
+			child_sections.emplace_back(child_start_idx, i - 1, key_section.depth + 1,
+			                            keys[i - 1]->data[key_section.depth]);
+			child_start_idx = i;
+		}
+	}
+	child_sections.emplace_back(child_start_idx, key_section.end, key_section.depth + 1,
+	                            keys[key_section.end]->data[key_section.depth]);
+}
+
+void Construct(vector<unique_ptr<Key>> &keys, row_t *row_ids, Node *&node, KeySection &key_section,
+               bool &has_constraint) {
+
+	auto &start_key = *keys[key_section.start];
```

Comment:
This could use an assertion making sure both `key_section.start` and `key_second.end` fall within the bounds of the `keys` vector